apollo-federation/src/sources/connect/json_selection/methods.rs

```diff
@@ -24,131 +21,161 @@ mod public;
-        methods.insert("not".to_string(), future::not_method);
-        methods.insert("or".to_string(), future::or_method);
-        methods.insert("and".to_string(), future::and_method);
+pub(super) trait ArrowMethodImpl {
+    fn apply(
+        &self,
+        method_name: &WithRange<String>,
+        method_args: Option<&MethodArgs>,
+        data: &JSON,
+        vars: &VarsWithPathsMap,
+        input_path: &InputPath<JSON>,
+        tail: &WithRange<PathList>,
+    ) -> (Option<JSON>, Vec<ApplyToError>);
+}
 
-        methods
-    };
+// This Deref implementation allows us to call .apply(...) directly on the
+// ArrowMethod enum.
+impl std::ops::Deref for ArrowMethod {
+    type Target = dyn ArrowMethodImpl;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            // Public methods:
+            Self::Echo => &public::EchoMethod,
+            Self::Map => &public::MapMethod,
+            Self::Match => &public::MatchMethod,
+            Self::First => &public::FirstMethod,
+            Self::Last => &public::LastMethod,
+            Self::Slice => &public::SliceMethod,
+            Self::Size => &public::SizeMethod,
+            Self::Entries => &public::EntriesMethod,
+
+            // Future methods:
+            Self::TypeOf => &future::TypeOfMethod,
+            Self::Eq => &future::EqMethod,
+            Self::MatchIf => &future::MatchIfMethod,
+            Self::Add => &future::AddMethod,
+            Self::Sub => &future::SubMethod,
+            Self::Mul => &future::MulMethod,
+            Self::Div => &future::DivMethod,
+            Self::Mod => &future::ModMethod,
+            Self::Has => &future::HasMethod,
+            Self::Get => &future::GetMethod,
+            Self::Keys => &future::KeysMethod,
+            Self::Values => &future::ValuesMethod,
+            Self::Not => &future::NotMethod,
+            Self::Or => &future::OrMethod,
+            Self::And => &future::AndMethod,
+        }
+    }
 }
 
-pub(super) fn lookup_arrow_method(method_name: &str) -> Option<&ArrowMethod> {
-    if cfg!(test) || PUBLIC_ARROW_METHODS.contains(method_name) {
-        ARROW_METHODS.get(method_name)
-    } else {
-        None
+impl ArrowMethod {
+    // This method is currently used at runtime to look up methods by &str name,
+    // but it could be hoisted parsing time, and then we'd store an ArrowMethod
+    // instead of a String for the method name in the AST.
+    pub(super) fn lookup(name: &str) -> Option<Self> {
+        let method_opt = match name {
+            "echo" => Some(Self::Echo),
+            "map" => Some(Self::Map),
+            "eq" => Some(Self::Eq),
+            "match" => Some(Self::Match),
+            // As this case suggests, we can't necessarily provide a name()
+            // method for ArrowMethod (the opposite of lookup), because method
+            // implementations can be used under multiple names.
+            "matchIf" | "match_if" => Some(Self::MatchIf),
```

Comment:
We could make the design decision _not_ to have multiple names, though. Seems like we should standardize on camelCase and not snake_case.

That would also mean we could use `strum` to impl both `FromStr` and `Display` for all the methods without all the boilerplate.

Reply:
I guess `strum` would make it harder to do the public/future split. Might end up having to add a `#[cfg(test)]` to every future method...

Reply:
I'm open to ideas, though not previously familiar with `strum`.

I'd be ok with sticking to one method name/style, but I think it's worthwhile to keep passing `method_name` as an argument into the method `apply`/`shape` implementations, so there's only one spot in the code we need to change if we want to rename a method (compared to each method implicitly "knowing" its own name and encoding it into error messages, for example).